### ISSUE TITLE
Custom Credential Provider for multistaged apps

### ISSUE THREAD

**Issue by BastianSperrhacke-Otto** [quarkusio/quarkus#8938]
**Description**
Hi guys. First of all: You are doing a good job, I like quarkus and the idea behind.
That's why I am currently playing with it and consider it as a new Techstack for out Microservices.

As I took one of our Java EE Apps and tried to migrate to quarkus I stumpled in a problem:
Currently it seems to be that the jpa connections and connection pools are configured with the application.properties. this file is usually checked in to git. So having multiple stages like live, test or prelive with different credentials (and jdbc urls) this is hard to realize with it. Additionally I think checking in a database's password in clear text is also not a good idea. 
As I googled around I found the vault extension for this issue which IMHO introduces the "credential Provider" pattern.
Unfortunately I do not have a vault server and also do not want to manage one. As we are in the AWS Cloud we fetch those credential at request time from AWS secretmanager or parameter store.

**Implementation ideas**
probably there is the chance to reuse the idea of an "credential Provider", but to give the chance to the user, to have an custom "credential Provider", which can be implemented by us.

If you have any other solutions for my issue plz share your idea.

Thank you and best regards.

Bastian

**Comment by geoand** [comment]
Hi,

Actually it's pretty easy to create your own "credentials provider" (or a more general "configuration provider"), but it does require you writing an extension (which however for what you need to do should be rather simple).

Here are two examples of such extensions:

https://github.com/quarkusio/quarkus/blob/77ce7dc8547b4279805956d8c886c5915a44c185/extensions/spring-cloud-config-client/pom.xml

https://github.com/quarkusio/quarkus/blob/feaddb819d1dd5dd0826c9fd1dcb3d4df6a46ad1/extensions/kubernetes-config/pom.xml

These extensions use a deeper integration into Quarkus that Vault uses (the Vault extension is older) so they are a lot simpler to write.

Although having integration points in regular application code (not an extension) does seem interesting, I am afraid that users could get into to trouble too easily and it would lead to hard to debug issues (as the code needs to run before other pieces and doesn't have access to all the usual Quarkus facilities).

**Comment by BastianSperrhacke-Otto** [comment]
Ok, looks promising. With these config sources (Microprofile, as it seems) it is also possible to inject at startup the correct jdbc connection parameters, right? I#ll give it a try next week.
Thank you.

**Comment by geoand** [comment]
Yes, any runtime configuration you could normally inject with `application.properties` (or env vars or java system props), you can inject using this method.

**Comment by sberyozkin** [comment]
If it is for Agroal then it is likely has to be an extension. This issue is probably a duplicate of  the combination of #8826 and #6896 (see comment from Vincent at https://github.com/quarkusio/quarkus/issues/6896#issuecomment-581014674)

**Comment by vsevel** [comment]
documentation now provided in #9552

**Comment by dasniko** [comment]
Hi guys,
I'm looking for something similar, reading all the DB information from an AWS secret, not only username/passwort, but also jdbc url information (which is a standard secret format in AWS Secrets Manager).

Would be very nice if there's a way to provide all this (jdbc) information easily.

@BastianSperrhacke-Otto Did you make any progress in the meantime building an own credentials provider?

**Comment by BastianSperrhacke-Otto** [comment]
@dasniko Hey Niko. We did not investigate so far. Sorry.

**Comment by vsevel** [comment]
hello @dasniko there is an official support for custom credentials provider and associated documentation.

**Comment by mvrk57** [comment]
Hi guys! any progress on getting aws secrets within quarkus? Im not sure if I should write my own config provider for that since it should be a usual use case? 
thanks a lot ;)
